### DISCUSSION TITLE
Tycho 2.7.5 & OpenJDK11 failed since 8 march 24

### DISCUSSION THREAD

**Discussion by StephFo** [eclipse-tycho/tycho#3573] [Q&A] [answered]
Hi Tycho Community,

I'm a tycho user for 7 years with no issue.
At the moment, my CI is based on OpenJDK Temurin 11.0.15 (April 2022) + Tycho 2.7.5.
Everything was fine until last 8 March 2024.

Indeed, Eclipse has built & populated maven central with a new version of bundle org.eclipse.core.runtime as 3.31.0 (https://central.sonatype.com/artifact/org.eclipse.platform/org.eclipse.core.runtime/versions) that seems to not be anymore compiled against Java11 (class file 55.0) but Java 17 (class file 61.0) instead of.

Tycho 2.7.5 depends on this bundle and seems not to depends on a specific version but a range.
I'm saying that because , running my build I can see all versions of org.eclipse.core.runtime bundle downloaded including 3.31.0 that breaks my build complaining about class file 61.0 which is not supported with java11.
These traces are before maven analyse my own bundles as modules...

Can a tycho committer confirm my assumption about a dependency range for org.eclipse.core.runtime  that is not strict enough ?

Does anyone have an idea to solve that issue please ? Would it be possible to get a 2.7.6 fixing the issue (excluding new eclipse bundle version compiled against Java 17 for 2.7.x branch) ?

Any help would be very much appreciated.

Migration towards Java17 JDK is one option but, I would like to know if an alternative is possible.

PS: Tycho 1.7.x builds are not impacted.

Regards,
Stephane.

**Comment by laeubi** [comment]
> Can a tycho committer confirm my assumption about a dependency range for org.eclipse.core.runtime that is not strict enough ?

I don't think Tycho itself is pulling this dependency, hut older releases of Eclipse has had dependencies that included items with an (open) version range what was quite bad as you notice here.

> Does anyone have an idea to solve that issue please ? Would it be possible to get a 2.7.6 fixing the issue (excluding new eclipse bundle version compiled against Java 17 for 2.7.x branch) ?

You can try to use `<exclude>` or explicit `<dependency>` elements on your Tycho plugin that pull ins the problematic version but that of course cumbersome, regarding older release please see https://github.com/eclipse-tycho/tycho/blob/master/CONTRIBUTING.md#backporting for details. 


To summarize, you either need to take actions to revive 2.7.x branch or you can pay someone to do the work for you but it is unlikely that current committers doing the work.

The reason is that Tycho 2.7.x is quite outdated, the current latest release is Tycho4, and you don't need to migrate your _code_ you just need to use a Java 17 JDK to *run* the maven build.

**Reply by AhmetBilgin** [reply]
Hi @laeubi,

I want to try fix the dependencies of tycho 2.7.5 (fix seems trivial) but I can't get a successful build even with unmodified sources.

I run the build with the following command (by looking into the JenkinsFile):
`mvn --batch-mode --global-settings maven-settings.xml --settings maven-settings.xml -U -V -e clean install -Pits -Dmaven.repo.local=$M2_REPO -DskipTests`

Tests failed and caused the build to fail. My build machine has no direct connection to the internet (just mirrors in our nexus) and some tests seem to access the url https://download.eclipse.org. So I added the `-DskipTests` option.
The build proceeds further, but then I run into a compilation error. I expected the original tycho source to be compiled successful.

I've attached the build logs.

- [mvn_with_tests.log](https://github.com/eclipse-tycho/tycho/files/14584290/mvn_with_tests.log)
- [mvn_skipTests.log](https://github.com/eclipse-tycho/tycho/files/14584286/mvn_skipTests.log)

Can you spot whats going wrong?

Thanks for the help

**Comment by AhmetBilgin** [comment]
Hi there,
we experience the exact same problem and are using the exact same versions of tycho 2.7.5 and jdk11.
None of our rcp developers can work currently.

That tycho has version ranges in it's dependencies is something we did not expect, as we are cautious with freezing version numbers of 3rd-party dependencies/plugins that we directly use.
...

**Comment by laeubi** [comment]
As said Tycho 2.x is EOL since 2022-09-30 without any community support/releases, if it is crucial to your business you can find dedicated support options for Tycho here:
https://github.com/eclipse-tycho/tycho#getting-support

**Reply by mickaelistria** [reply, marked as answer]
Note that while I don't want to dissuade anyone from buying support, I would personally suggest anyone willing to invest support to "keep old versions" to invest the same effort in getting support for "adopt new versions", it will be much more profitable to everyone.
As mentioned already, even if Maven/Tycho uses Java 17, it can still build and target code for older Java versions. Basically, just install Java 17 as non-default, prefix you `mvn` command with `JAVA_HOME=/path/to/java17 ` so the Java 17 installation is found and your next Maven commands will succeed with newer Tycho, without changing your PATH or anything.

**Reply by laeubi** [reply]
The same would maybe work for Tycho 2.7.5, but bottom line is its out of support, so even if one should plan for upgrade to not be surprised by something else, e.g Tycho 2.x can probably not read latest platform releases because of changed hash-sums or usage of PGP or whatever else could happen.

**Reply by StephFo** [reply]
Thanks for the tips.
To workaround the issue , I now use OpenJDK17 (Temurin) set as JAVA_HOME but still targeting java11 in compilation definition.
As explained above.

I did not have to modify any source code so far.
Generated class file are class file 55.0 as before. RCP Application runs smoothly with initial OpenJDK11 JRE.
See pom file example below.


I understand the position of Tyco committers but it is strange to see that rcp builds with Tycho 1.7 & Java8 JRE still run fine whereas Tycho 2.7.5 quite recent (september 2022, only 18 months ago) is broken.

Thank you for the support anyway.

pom file example:

<properties>
		<javase.version>JavaSE-11</javase.version>
		<java.version>11</java.version>
</properties>

<plugin>
	       <groupId>org.eclipse.tycho</groupId>
		<artifactId>target-platform-configuration</artifactId>
		<version>${tycho-version}</version>
		<configuration>
			**<executionEnvironment>${javase.version}</executionEnvironment>**
...
</plugin>

<plugin>
		<groupId>org.eclipse.tycho</groupId>
		<artifactId>tycho-compiler-plugin</artifactId>
		<version>${tycho-version}</version>
		<configuration>
		<encoding>UTF-8</encoding>
		<useProjectSettings>false</useProjectSettings>
		**<release>${java.version}</release>**
...
</plugin>

**Reply by laeubi** [reply]
The problem is that we suffer from "bad" platform versions policy here, this was fixed in Tycho 3 / 4 now as we had this problems in the past, so one does not get "surprising" updates anymore.

**Reply by StephFo** [reply]
Good to know Thanks

**Reply by AhmetBilgin** [reply]
Hi everyone,
thanks for the quick help.
Using JDK17 set as JAVA_HOME also worked for us, although we had to switch of the parallel build. Otherwise  we ran into race conditions during the build.

@StephFo do you run the build in parallel?

**Reply by laeubi** [reply]
> Using JDK17 set as JAVA_HOME also worked for us, although we had to switch of the parallel build. Otherwise we ran into race conditions during the build.

That should be fixed in later ecj releases (included in Tycho 3+).

**Comment by rgra** [comment]
You can see where the offending versions are pulled in by looking at the tree for org.eclipse.tycho:tycho-build:jar:2.7.0.
See https://github.com/eclipse-equinox/p2/issues/481#issue-2183739537

I can not run the build with Java 17 because the java.xml dependencies are missing in that version.

I modified the poms of 
```
org.eclipse.equinox.p2.core-2.5.100.pom
org.eclipse.equinox.p2.director-2.5.100.pom
org.eclipse.equinox.p2.engine-2.7.200.pom
```
to exclude the new versions

>      <dependency>
>         <artifactId>org.eclipse.equinox.common</artifactId>
>         <version>[3.5.0,3.19.0)</version>
>     </dependency>
>      <dependency>
>         <groupId>org.eclipse.platform</groupId>
>         <artifactId>org.eclipse.core.jobs</artifactId>
>         <version>[3.3.0,3.15.0)</version>
>     </dependency>
>         <dependency>
>       <groupId>org.eclipse.platform</groupId>
>       <artifactId>org.eclipse.equinox.registry</artifactId>
>       <version>[0.0,3.12.0)</version>
>     </dependency>

As we are running the builds via our own nexus repositories I uploaded the modified poms and respective jars to our releases repository.

The builds now run normally as before.

**Reply by StephFo** [reply]
With Tycho 2.7.5 +  Temurin OpenJDK 17 no issue. 
Not sure to understand what you spotted about **java.xml** dependencies are missing in that version.
It seems you are using Java17 to run maven + Tycho and compile against, aren't you ?
You can run maven with java17 but still compiles against Java11 as explained.
Even with only OpenJDK 11, you should setup ExecutionEnvironment and  compiler target as shown below.
Like this if you run maven with OpenJDK17, generated classes are still Java11 compatible ones.

```
<properties>
		<javase.version>JavaSE-11</javase.version>
		<java.version>11</java.version>
</properties>

<plugin>
	       <groupId>org.eclipse.tycho</groupId>
		<artifactId>target-platform-configuration</artifactId>
		<version>${tycho-version}</version>
		<configuration>
			<executionEnvironment>${javase.version}</executionEnvironment>
...
</plugin>

<plugin>
		<groupId>org.eclipse.tycho</groupId>
		<artifactId>tycho-compiler-plugin</artifactId>
		<version>${tycho-version}</version>
		<configuration>
		<encoding>UTF-8</encoding>
		<useProjectSettings>false</useProjectSettings>
		<release>${java.version}</release>
...
</plugin>

```

My apologies I've badly copy/paste in my initial post.

**Reply by raymond-pg** [reply]
Is there a way to control the version of org.eclipse.equinox.common? I tried to set dependency in my pom, but it is not working.
Or is there a way to use my own repository instead of official maven one?
